gpsd/man/alfred-gpsd.8
.\" SPDX-License-Identifier: GPL-2.0
.\" License-Filename: LICENSES/preferred/GPL-2.0
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH "ALFRED-GPSD" "8" "Oct 04, 2013" "Linux" "Alfred GPS distribution server"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.\" --------------------------------------------------------------------------
.\" Process this file with
.\" groff -man batadv-vis.8 -Tutf8
.\" Retrieve format warnings with
.\" man --warnings batadv-vis.8 > /dev/null
.\" --------------------------------------------------------------------------
.ad l
.SH NAME
alfred\-gpsd \- Alfred GPS distribution server
.SH SYNOPSIS
.B alfred\-gpsd [\fIoptions\fP]
.br
.SH DESCRIPTION
alfred\-gpsd can be used to distribute GPS location information over
your batman-adv mesh network. It reads the current location from gpsd,
or a fixed location from the command line, and distributes this
information via alfred. By gathering this local information, any
alfred-gpsd node can get location information about other nodes.
.PP
.PP
.SH OPTIONS
.TP
\fB\-v\fP, \fB\-\-version\fP
Print the version
.TP
\fB\-h\fP, \fB\-\-help\fP
Display a brief help message.
.TP
\fB\-u\fP, \fB\-\-unix-path\fP \fIpath\fP
path to unix socket used for alfred server communication.
.TP
\fB\-s\fP, \fB\-\-server\fP
Start up in server mode. This server will read the current location
from gpsd and set it in alfred via unix socket. The alfred server must
run too to get this information set.
.TP
\fB\-l\fP, \fB\-\-location <lat>,<lon>,<alt>\fP
Rather than read the current location from gpsd, use a fixed location.
.TP
\fB\-g\fP, \fB\-\-gpsd server[:port[:device]]\fP
Specify the server hostname and optional port of where gpsd is
listening. Additionally, a specific device connected to gpsd can be specified.
.
.SH EXAMPLES
Start an alfred\-gpsd server which is fetching GPS data from the local gpsd.
.br
\fB     alfred\-gpsd \-s\fP
.br

Start an alfred\-gpsd server with a fixed location
.br
\fB     alfred\-gpsd \-s \-l 48.858222,2.2945,358\fP
.br

To get a list of GPS locations, in JSON format:
.br
\fB     alfred-gpsd\fP
.nf
[
  { "source" : "f6:00:48:13:d3:1e", "tpv" : {"class":"TPV","tag":"RMC",
    "device":"/dev/ttyACM0","mode":3,"time":"2013-10-01T10:43:20.000Z",
    "ept":0.005,"lat":52.575485000,"lon":-1.339716667,"alt":122.500,
    "epx":10.199,"epy":15.720,"epv":31.050,"track":0.0000,"speed":0.010,
    "climb":0.000,"eps":31.44} },
  { "source" : "8e:4c:77:b3:65:b4", "tpv" : {"class":"TPV",
    "device":"command line","time":"2013-10-01T10:43:05.129Z",
    "lat":48.858222,"lon":2.2945,"alt":358.000000,"mode":3}
  }
]
.fi
.br
.
.SH SEE ALSO
.BR alfred (8),
.BR gpsd (8)
.BR gpsd_json(5)
.SH AUTHOR
alfred\-gpsd and this Manual page was written by Andrew Lunn <andrew@lunn.ch>


CHANGELOG
alfred 2017.3:

 * reduction of memory usage when using --update-command

 -- Thu, 28 Sep 2017 13:43:18 +0200

alfred 2017.2:

 * reduce cpu load when rating multiple master servers
 * coding style cleanups and refactoring

 -- Fri, 28 Jul 2017 13:45:50 +0200

alfred 2017.1:

 * (no changes)

 -- Tue, 23 May 2017 11:40:21 +0200

alfred 2017.0:

 * support IPv4 multicast distribution
 * coding style cleanups

 -- Tue, 28 Feb 2017 17:22:49 +0100

alfred 2016.5:

 * support interface validity checks on systems without debugfs
 * remove debugfs check during batadv-vis startup
 * allow out-of-order txend packets during transmissions

 -- Thu, 15 Dec 2016 18:11:55 +0100

alfred 2016.4:

 * add expert option to specify sync interval
 * fix various bugs in batadv-vis netlink integration
 * fix build build problems with libnl-tiny

 -- Thu, 27 Oct 2016 21:22:32 +0200

alfred 2016.3:

 * integrate support for batman-adv netlink

 -- Thu, 01 Sep 2016 16:18:35 +0200

alfred 2016.2:

 * add support for automatic debugfs mount with enabled 
   CONFIG_ALFRED_CAPABILITIES

 -- Thu, 09 Jun 2016 16:21:19 +0200

alfred 2016.1:

 * add support for master servers to receive push_data packets with foreign
   source addresses
 * various code cleanups
 * bugs squashed:
   - ignore invalid EUI64 addresses

 -- Thu, 21 Apr 2016 14:09:12 +0200

alfred 2016.0:

 * various code and documentation cleanups

 -- Tue, 19 Jan 2016 16:44:26 +0100

alfred 2015.2:

 * mention libcap in the README
 * Fix typos

 -- Mon, 23 Nov 2015 20:35:20 +0100

alfred 2015.1:

 * add support to run on interfaces with multiple link-local addresses
 * various code cleanups
 * bugs squashed:
   - reduce of maximum payload size to always fit into UDP datagrams

 -- Tue, 04 Aug 2015 14:59:10 +0200

alfred 2015.0:

 * add support to call commands after data was updated
 * automatic reduction of process capabilities when not needed anymore
 * allow printing of the data version number in the alfred client mode
 * various code cleanups
 * bugs squashed:
   - update of the version number when data was updated with different
     version number
   - tighten size check on received packet

 -- Tue, 28 Apr 2015 20:39:51 +0200

alfred 2014.4.0:

 * add support for multiple interfaces per master
 * add support for changing interfaces on the fly
 * changes to support multiple alfred interfaces:
   - bind alfred to a specific interface
   - allow configuring the unix socket path
 * enhanced debugging

 -- Wed, 31 Dec 2014 00:40:42 +0100

alfred 2014.3.0:

 * fix various possible memleak, access errors and strncpy issues
 * handle fcntl return codes
 * fix altitude verification check in gpsd

 -- Mon, 21 Jul 2014 16:53:01 +0200

alfred 2014.2.0:

 * Handle EPERM errors on every sendto
 * Check for changed interface properties, e.g. recreation or
   changed MAC- and IPv6 addresses

 -- Thu, 15 May 2014 19:41:44 +0200

alfred 2014.1.0:

 * don't leak socket fd in batadv-vis

 -- Thu, 13 Mar 2014 12:10:26 +0100

alfred 2014.0.0:

 * add installation of the alfred-gpsd manpage
 * add -lm to linker flags for libgps in alfred-gpsd

 -- Sat, 04 Jan 2014 20:17:07 +0100

alfred 2013.4.0:

 * add new json output format for vis
 * add gps location information service for alfred
 * allow network interface to vanish and return without restart
 * allow to switch between master and slave operation without restart
 * renamed vis to batadv-vis to avoid collisions with other vis binaries
 * add manpages
 * various code cleanups
 * bugs squashed:

   * handle failing write() in unix sockets
   * Fix crash when vis opened empty file

 -- Sun, 13 Oct 2013 22:55:43 +0200

alfred 2013.3.0:

 * initial release of alfred after beta (synced release cycle with
   batman-adv)
 * allows to share arbitrary local information over a (mesh) network
 * initial support for vis (previously in-kernel feature of batman-adv
   to visualize the network) included
 * easy but flexible communication interface to allow data applications
   of all kinds
 * two-tiered architecture (master and slaves)
 * exchanges data via IPv6 unicast/multicast

 -- Sun, 28 Jul 2013 18:56:52 +0200


gpsd/alfred-gpsd.c
// SPDX-License-Identifier: GPL-2.0
/* Copyright (C) 2013-2017  B.A.T.M.A.N. contributors:
 *
 * Andrew Lunn, Simon Wunderlich
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 *
 * License-Filename: LICENSES/preferred/GPL-2.0
 */

#include "alfred-gpsd.h"

static struct globals gpsd_globals;

static int alfred_open_sock(struct globals *globals)
{
	struct sockaddr_un addr;

	globals->unix_sock = socket(AF_LOCAL, SOCK_STREAM, 0);
	if (globals->unix_sock < 0) {
		perror("can't create unix socket");
		return -1;
	}

	memset(&addr, 0, sizeof(addr));
	addr.sun_family = AF_LOCAL;
	strncpy(addr.sun_path, globals->unix_path, sizeof(addr.sun_path));
	addr.sun_path[sizeof(addr.sun_path) - 1] = '\0';

	if (connect(globals->unix_sock, (struct sockaddr *)&addr,
		    sizeof(addr)) < 0) {
		close(globals->unix_sock);
		globals->unix_sock = -1;
		perror("can't connect to unix socket");
		return -1;
	}

	return 0;
}

static int gpsd_publish_data(struct globals *globals)
{
	int len, ret;

	/* to push data we have to add a push header, the header for the data
	 * and our own data type.
	 */
	globals->push->tx.id = htons(ntohs(globals->push->tx.id) + 1);

	len = GPSD_DATA_SIZE(globals->gpsd_data);
	globals->push->data->header.length = htons(len);
	len += sizeof(*globals->push) - sizeof(globals->push->header);
	len += sizeof(*globals->push->data);
	globals->push->header.length = htons(len);
	len +=  sizeof(globals->push->header);

	alfred_open_sock(globals);
	if (globals->unix_sock < 0)
		return globals->unix_sock;

	ret = write(globals->unix_sock, globals->buf, len);
	close(globals->unix_sock);
	if (ret < len)
		return -1;

	return 0;
}

static void gpsd_get_location(struct globals *globals)
{
	if (globals->source == SOURCE_CMDLINE) {
		char tbuf[JSON_DATE_MAX+1];
		timestamp_t now = timestamp();

		sprintf(globals->gpsd_data->tpv,
			"{\"class\":\"TPV\",\"device\":\"command line\","
			"\"time\":\"%s\","
			"\"lat\":%f,\"lon\":%f,\"alt\":%f,"
			"\"mode\":3}",
			unix_to_iso8601(now, tbuf, sizeof(tbuf)),
			globals->lat, globals->lon, globals->alt);
		globals->gpsd_data->tpv_len =
			htonl(strlen(globals->gpsd_data->tpv) + 1);
	}
}

static int gpsd_update_data(struct globals *globals)
{
	gpsd_get_location(globals);
	gpsd_publish_data(globals);

	return 0;
}

static int gpsd_request_data(struct globals *globals)
{
	int ret;

	globals->request = (struct alfred_request_v0 *) globals->buf;

	globals->request->header.type = ALFRED_REQUEST;
	globals->request->header.version = ALFRED_VERSION;
	globals->request->header.length =
		htons(sizeof(*globals->request) - 
		      sizeof(globals->request->header));
	globals->request->requested_type = GPSD_PACKETTYPE;
	globals->request->tx_id = htons(random());

	alfred_open_sock(globals);
	if (globals->unix_sock < 0)
		return globals->unix_sock;

	ret = write(globals->unix_sock, globals->request,
		    sizeof(*globals->request));
	if (ret < (int)sizeof(*globals->request)) {
		close(globals->unix_sock);
		return -1;
	}

	return globals->unix_sock;
}

static struct gpsd_v1 *gpsd_receive_answer_packet(int sock, uint16_t *len,
						  uint8_t *source)
{
	static uint8_t buf[65536];
	struct alfred_tlv *tlv;
	struct alfred_push_data_v0 *push;
	struct alfred_data *data;
	int l, ret;

	ret = read(sock, buf, sizeof(*tlv));
	if (ret < 0)
		return NULL;

	if (ret < (int)sizeof(*tlv))
		return NULL;

	tlv = (struct alfred_tlv *)buf;
	/* TODO: might return an ALFRED_STATUS_ERROR too, handle it */
	if (tlv->type != ALFRED_PUSH_DATA)
		return NULL;

	l = ntohs(tlv->length);
	/* exceed the buffer? don't read */
	if (l > (int)(sizeof(buf) - sizeof(push->header)))
		return NULL;

	/* not enough for even the push packet and header? don't bother. */
	if (l < (int)(sizeof(*push) - sizeof(push->header) + sizeof(*data)))
		return NULL;

	/* read the rest of the packet */
	ret = read(sock, buf + sizeof(*tlv), l);
	if (ret < l)
		return NULL;

	push = (struct alfred_push_data_v0 *)buf;
	data = push->data;
	*len = ntohs(data->header.length);

	if (data->header.type != GPSD_PACKETTYPE)
		return NULL;

	if (data->header.version != GPSD_PACKETVERSION)
		return NULL;

	memcpy(source, data->source, ETH_ALEN);
	return (struct gpsd_v1 *) data->data;
}

static int gpsd_read_answer(struct globals *globals)
{
	struct gpsd_v1 *gpsd_data;
	uint16_t len;
	uint8_t source[ETH_ALEN];
	bool first_line = true;

	printf("[\n");

	while ((gpsd_data = gpsd_receive_answer_packet(globals->unix_sock,
						       &len,
						       source)) != NULL) {
		if (len < sizeof(*gpsd_data))
			break;

		/* check size and skip bogus packets */
		if (len != GPSD_DATA_SIZE(gpsd_data))
			continue;

		if (first_line)
			first_line = false;
		else
			printf(",\n");

		printf("  { \"source\" : \"%02x:%02x:%02x:%02x:%02x:%02x\", "
		       "\"tpv\" : %s }",
		       source[0], source[1], source[2],
		       source[3], source[4], source[5],
		       gpsd_data->tpv);
	}
	printf("\n]\n");


	return 0;
}

/* Standard parsing of a GPS data source spec. Taken from gpsdclient.c */
static void gpsd_source_spec(const char *arg, struct fixsource_t *source)
{
	/* the casts attempt to head off a -Wwrite-strings warning */
	source->server = (char *)"localhost";
	source->port = (char *)DEFAULT_GPSD_PORT;
	source->device = NULL;

	if (arg != NULL) {
		char *colon1, *skipto, *rbrk;
		source->spec = strdup(arg);
		assert(source->spec != NULL);

		skipto = source->spec;
		if (*skipto == '[' && (rbrk = strchr(skipto, ']')) != NULL) {
			skipto = rbrk;
		}
		colon1 = strchr(skipto, ':');

		if (colon1 != NULL) {
			char *colon2;
			*colon1 = '\0';
			if (colon1 != source->spec) {
				source->server = source->spec;
			}
			source->port = colon1 + 1;
			colon2 = strchr(source->port, ':');
			if (colon2 != NULL) {
				*colon2 = '\0';
				source->device = colon2 + 1;
			}
		} else if (strchr(source->spec, '/') != NULL) {
			source->device = source->spec;
		} else {
			source->server = source->spec;
		}
	}

	if (*source->server == '[') {
		char *rbrk = strchr(source->server, ']');
		++source->server;
		if (rbrk != NULL)
			*rbrk = '\0';
	}
}

static int gpsd_get_data(struct globals *globals)
{
	globals->unix_sock = gpsd_request_data(globals);
	if (globals->unix_sock < 0)
		return -1;

	gpsd_read_answer(globals);
	close(globals->unix_sock);

	return 0;
}

static void gpsd_connect_gpsd(struct globals *globals)
{
	unsigned int flags = WATCH_ENABLE | WATCH_JSON;
	int ret;

	ret = gps_open(globals->gpsdsource.server, globals->gpsdsource.port,
		       &globals->gpsdata);

	if (ret) {
		/* Could not connect to gpsd. Set the fd so we don't
		   try to perform select(2) on it. */
		globals->gpsdata.gps_fd = -1;
		return;
	}

	if (globals->gpsdsource.device != NULL)
		flags |= WATCH_DEVICE;

	gps_stream(&globals->gpsdata, flags, globals->gpsdsource.device);
}

static void gpsd_read_gpsd(struct globals *globals)
{
	ssize_t ret;
	size_t cnt;
	bool eol = false;
	char buf[4096];
	const size_t tpv_size = sizeof(globals->buf) -
				sizeof(*globals->push) -
				sizeof(struct alfred_data) -
				sizeof(*globals->gpsd_data);

	cnt = 0;
	do {
		ret = read(globals->gpsdata.gps_fd, &buf[cnt], 1);
		if (ret != 1) {
			gps_close(&globals->gpsdata);
			globals->gpsdata.gps_fd = -1;
			return;
		}

		switch (buf[cnt]) {
		case '\r':
			cnt--;
			break;
		case '\n':
			eol = true;
			buf[cnt] = '\0';
			break;
		}
	} while (cnt++ < sizeof(buf) - 1 && !eol);

	if (!eol) {
		gps_close(&globals->gpsdata);
		globals->gpsdata.gps_fd = -1;
		return;
	}

#define STARTSWITH(str, prefix)	strncmp(str, prefix, sizeof(prefix)-1)==0
	if (STARTSWITH(buf, "{\"class\":\"TPV\"")) {
		strncpy(globals->gpsd_data->tpv, buf, tpv_size);
		globals->gpsd_data->tpv[tpv_size - 1] = '\0';

		globals->gpsd_data->tpv_len =
			htonl(strlen(globals->gpsd_data->tpv) + 1);
	}
}

static void gpsd_usage(void)
{
	printf("Usage: alfred-gpsd [options]\n");
	printf("  -s, --server                start up in server mode, which regularly updates gpsd data from batman-adv\n");
	printf("  -l <lat>,<lon>,<alt>        Static location\n");
	printf("  -g server[:port[:device]]   GPSD source\n");
	printf("  -v, --version               print the version\n");
	printf("  -h, --help                  this help\n");
	printf("\n");
}

static void gpsd_parse_location(struct globals *globals,
				const char * optarg)
{
	int n;
	float lat, lon, alt;

	n = sscanf(optarg, "%f,%f,%f", &lat, &lon, &alt);
	if (n != 3) {
		printf("Unable to parse location\n");
		gpsd_usage();
		exit(EXIT_FAILURE);
	}

	if ((lat < -90) || (lat > 90)) {
		printf("Invalid latitude\n");
		gpsd_usage();
		exit(EXIT_FAILURE);
	}

	if ((lon < -180) || (lon > 180)) {
		printf("Invalid longitude\n");
		gpsd_usage();
		exit(EXIT_FAILURE);
	}

	if ((alt < -1000) || (alt > 9000)) {
		/* No support for aircraft or submarines! */
		printf("Invalid altitude\n");
		gpsd_usage();
		exit(EXIT_FAILURE);
	}

	globals->lat = lat;
	globals->lon = lon;
	globals->alt = alt;
}

static struct globals *gpsd_init(int argc, char *argv[])
{
	bool have_source = false;
	int opt, opt_ind;
	struct globals *globals;
	struct option long_options[] = {
		{"server",	no_argument,		NULL,	's'},
		{"location",    required_argument,	NULL,	'l'},
		{"gpsd",	required_argument,	NULL,	'g'},
		{"unix-path", 	required_argument,	NULL,	'u'},
		{"help",	no_argument,		NULL,	'h'},
		{"version",	no_argument,		NULL,	'v'},
		{NULL,		0,			NULL,	0},
	};

	globals = &gpsd_globals;
	memset(globals, 0, sizeof(*globals));

	globals->opmode = OPMODE_CLIENT;
	globals->source = SOURCE_GPSD;
	globals->gpsd_format = FORMAT_JSON;
	globals->unix_path = ALFRED_SOCK_PATH_DEFAULT;

	while ((opt = getopt_long(argc, argv, "shl:g:vu:", long_options,
				  &opt_ind)) != -1) {
		switch (opt) {
		case 's':
			globals->opmode = OPMODE_SERVER;
			break;
		case 'l':
			globals->source = SOURCE_CMDLINE;
			gpsd_parse_location(globals, optarg);
			break;
		case 'g':
			gpsd_source_spec(optarg, &globals->gpsdsource);
			have_source = true;
			break;
		case 'u':
			globals->unix_path = optarg;
			break;
		case 'v':
			printf("%s %s\n", argv[0], SOURCE_VERSION);
			printf("GPSD alfred client\n");
			return NULL;
		case 'h':
		default:
			gpsd_usage();
			return NULL;
		}
	}

	if (globals->source == SOURCE_GPSD && !have_source)
		gpsd_source_spec(NULL, &globals->gpsdsource);

	if (signal(SIGPIPE, SIG_IGN) == SIG_ERR)
		perror("could not register SIGPIPE handler");
	return globals;
}

static int gpsd_server(struct globals *globals)
{
	struct timeval tv;
	fd_set fds;
	int max_fd, ret;
	const size_t overhead = sizeof(*globals->push) +
		sizeof(struct alfred_data);
	const size_t tpv_size = sizeof(globals->buf) -
				sizeof(*globals->push) -
				sizeof(struct alfred_data) -
				sizeof(*globals->gpsd_data);
	long interval;

	globals->push = (struct alfred_push_data_v0 *) globals->buf;
	globals->gpsd_data = (struct gpsd_v1 *)
		(globals->buf + overhead);

	globals->push->header.type = ALFRED_PUSH_DATA;
	globals->push->header.version = ALFRED_VERSION;
	globals->push->tx.id = 0;
	globals->push->tx.seqno = 0;
	globals->push->data->header.type = GPSD_PACKETTYPE;
	globals->push->data->header.version = GPSD_PACKETVERSION;

	strncpy(globals->gpsd_data->tpv, GPSD_INIT_TPV, tpv_size);
	globals->gpsd_data->tpv[tpv_size - 1] = '\0';
	globals->gpsd_data->tpv_len =
		htonl(strlen(globals->gpsd_data->tpv) + 1);

	/* If we have a static location, we don't need to update very
	   often. */
	if (globals->source == SOURCE_GPSD) {
		globals->gpsdata.gps_fd = -1;
		interval = 2;
	} else
		interval = 60 * 5;

	while (1) {
		gpsd_update_data(globals);

		/* If we are not connected to gpsd, try to connect. */
		if (globals->source == SOURCE_GPSD &&
		    globals->gpsdata.gps_fd == -1) {
			gpsd_connect_gpsd(globals);
		}

		/* Use linux's select(2) behaviour of setting
		   tv to the remaining time when it exists */
		tv.tv_sec = interval;
		tv.tv_usec = 0;

		do {
			FD_ZERO(&fds);

			if (globals->source == SOURCE_GPSD &&
				globals->gpsdata.gps_fd != -1) {
				FD_SET(globals->gpsdata.gps_fd, &fds);
				max_fd = globals->gpsdata.gps_fd + 1;
			} else {
				max_fd = 0;
			}

			errno = 0;
			ret = select(max_fd, &fds, NULL, NULL, &tv);
			if (ret == -1 && errno != EINTR)
				perror("select error");

			if (ret == 1)
				gpsd_read_gpsd(globals);
		} while (ret != 0);
	}
	return EXIT_FAILURE;
}

int main(int argc, char *argv[])
{
	struct globals *globals;

	globals = gpsd_init(argc, argv);

	if (!globals)
		return EXIT_FAILURE;

	switch (globals->opmode) {
	case OPMODE_SERVER:
		return gpsd_server(globals);
		break;
	case OPMODE_CLIENT:
		return gpsd_get_data(globals);
		break;
	}

	return EXIT_FAILURE;
}
